cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll

```diff
@@ -5,10 +5,458 @@
+  class ArgumentPosition extends int {
+    ArgumentPosition() { this = parameterPosition() }
+  }
+
+  /** Holds if arguments at position `apos` match parameters at position `ppos`. */
+  overlay[caller?]
+  pragma[inline]
+  predicate parameterMatch(ParameterPosition ppos, ArgumentPosition apos) { ppos = apos }
+
+  final private class FinalMethod = Cpp::Function;
+
+  /**
+   * A non-overridable function.
+   *
+   * This function is non-overrideable either because it is not a member function, or
+   * because it is a final member function.
+   */
+  class NonOverridableMethod extends FinalMethod {
+    NonOverridableMethod() {
+      not this instanceof Cpp::MemberFunction
+      or
+      exists(Cpp::MemberFunction mf | this = mf |
+        not mf.isVirtual()
+        or
+        mf.isFinal()
+      )
+    }
+
+    /** Gets the `Parameter` at `pos` of this function, if any. */
+    Parameter getParameter(ParameterPosition ppos) { super.getParameter(ppos) = result }
+
+    /** Gets an expression returned from this function. */
+    GuardsInput::Expr getAReturnExpr() {
+      exists(ReturnValueInstruction ret |
+        ret.getEnclosingFunction() = this and
+        result = ret.getReturnValue()
+      )
+    }
+  }
+
+  private predicate nonOverridableMethodCall(CallInstruction call, NonOverridableMethod m) {
+    call.getStaticCallTarget() = m
+  }
+
+  /**
+   * A call to a `NonOverridableMethod`.
+   */
+  class NonOverridableMethodCall extends GuardsInput::Expr instanceof CallInstruction {
+    NonOverridableMethodCall() { nonOverridableMethodCall(this, _) }
+
+    /** Gets the function that is called. */
+    NonOverridableMethod getMethod() { nonOverridableMethodCall(this, result) }
+
+    /** Gets the argument at `apos`, if any. */
+    GuardsInput::Expr getArgument(ArgumentPosition apos) { result = super.getArgument(apos) }
+  }
+}
+
+private module GuardsImpl = SharedGuards::Make<Cpp::Location, IRCfg, GuardsInput>;
+
+private module LogicInput_v1 implements GuardsImpl::LogicInputSig {
+  private import semmle.code.cpp.dataflow.new.DataFlow::DataFlow::Ssa
+
+  final private class FinalBaseSsaVariable = Definition;
+
+  class SsaDefinition extends FinalBaseSsaVariable {
+    GuardsInput::Expr getARead() { result = this.getAUse().getDef() }
+  }
+
+  class SsaWriteDefinition extends SsaDefinition instanceof ExplicitDefinition {
+    GuardsInput::Expr getDefinition() { result = super.getAssignedInstruction() }
+  }
+
+  class SsaPhiNode extends SsaDefinition instanceof PhiNode {
+    predicate hasInputFromBlock(SsaDefinition inp, BasicBlock bb) {
+      super.hasInputFromBlock(inp, bb)
+    }
+  }
+
+  predicate parameterDefinition(GuardsInput::Parameter p, SsaDefinition def) {
+    def.isParameterDefinition(p)
+  }
+
+  predicate additionalImpliesStep(
+    GuardsImpl::PreGuard g1, GuardValue v1, GuardsImpl::PreGuard g2, GuardValue v2
+  ) {
+    g1.(ConditionalBranchInstruction).getCondition() = g2 and
```

Comment:
Should this perhaps be an `IdExpr` instead of an `additionalImpliesStep`? I think that amounts to mostly the same, right?

Reply:
Done in 99e1a07b8e6a0a471a2bcfd55124588957da20d3